Holovibes/cuda_sources/bw_area.cu

```diff
@@ -7,188 +7,205 @@
-
-        if (!nn)
+        if (min_l < l)
         {
-            const bool nne = ((y > 0) && (x > 0)) ? (value == image_d[(x - 1) * width + y - 1]) : false;
-            const bool ne = (x > 0) ? (value == image_d[(x - 1) * width + y]) : false;
-            const bool nw = ((y > 0) && (x < width - 1)) ? (value == image_d[(x + 1) * width + y - 1]) : false;
-
-            if (nw)
-            {
-                if ((nne && ne) || (nne && !ne))
-                    reduction(labels_d, labels_d[x * width + y], labels_d[(x + 1) * width + y - 1]);
-
-                if (!nne && ne)
-                    reduction(labels_d, labels_d[x * width + y], labels_d[(x - 1) * width + y]);
-            }
+            labels[idx] = min_l;
+            *change = 1;
         }
     }
 }
 
-__global__ void resolve_background(uint* labels_d, const float* image_d, float* labels_sizes_d, const size_t size)
+__global__ void update_label_kernel(uint* labels, uint* linked, const size_t size)
 {
-    int idx = blockIdx.x * blockDim.x + threadIdx.x;
+    size_t idx = threadIdx.x + blockIdx.x * blockDim.x;
 
     if (idx < size)
     {
-        labels_d[idx] = (!IS_BACKGROUND(image_d[idx])) ? labels_d[idx] + 1 : 0;
-        if (labels_d[idx])
-            atomicAdd(labels_sizes_d + labels_d[idx], 1.0f);
+        uint l = labels[idx];
+        if (l != linked[idx])
+            labels[idx] = find(linked, l);
     }
 }
 
 void get_connected_component(uint* labels_d,
-                             float* labels_sizes_d,
                              uint* linked_d,
                              const float* image_d,
                              const size_t width,
                              const size_t height,
+                             size_t* change_d,
                              const cudaStream_t stream)
 {
-    size_t size = width * height;
+    // Setting up the parallelisation.
+    uint threads_2d = get_max_threads_2d();
+    dim3 lthreads(threads_2d, threads_2d);
+    dim3 lblocks(1 + (width - 1) / threads_2d, 1 + (height - 1) / threads_2d);
+
     uint threads = get_max_threads_1d();
-    uint blocks = map_blocks_to_problem(size, threads);
+    uint blocks = map_blocks_to_problem(width * height, threads);
+
+    initialisation_kernel<<<lblocks, lthreads, 0, stream>>>(image_d, labels_d, linked_d, height, width);
+    size_t change_h;
+    do
+    {
+        cudaXMemset(change_d, 0, sizeof(size_t));
+
+        propagate_labels_kernel<<<lblocks, lthreads, 0, stream>>>(labels_d, linked_d, height, width, change_d);
+        cudaCheckError();
+        cudaXStreamSynchronize(stream);
+        cudaXMemcpy(&change_h, change_d, sizeof(size_t), cudaMemcpyDeviceToHost);
+
+        if (change_h)
+        {
+            update_label_kernel<<<blocks, threads, 0, stream>>>(labels_d, linked_d, width * height);
+            cudaCheckError();
+        }
+
+    } while (change_h);
+    cudaXStreamSynchronize(stream);
```

Comment:
You can probably use async versions of cudaXMemset and cudaXMemcpy using the same stream you use in the kernels to avoid having to cudaXStreamSynchronize, the order will be respected as stream actions act as a queue